### DISCUSSION TITLE
Extending Base but optional _id?

### DISCUSSION THREAD

**Discussion by theDanielJLewis** [typegoose/typegoose#851] [Q&A]
I'm really confused and frustrated and I hope you can help!

If I extend `Base`, I can easily call `_id` or `id`, but any time I create a new document, TypeScript complains that `_id` is required!

If I try to make `_id` optional, I get the error `Interface 'MyClass' incorrectly extends interface 'Base<ObjectId>'.`

But if I don't use `Base` and instead add the following to MyClass:

```js
	@prop()
	public _id?: Types.ObjectId;

	@prop()
	public id?: string;
```

Then the `id` value is always empty.

But if I remove the `id` part, then I get errors that it doesn't exist when I try to use it instead of `_id.toString()`!

GAH!

Please help!

Thanks in advance!

**Comment by hasezoey** [comment]
> But if I don't use Base and instead add the following to MyClass:
> `@prop()`

read [I need to set _id manually, how to do it properly?](https://typegoose.github.io/typegoose/docs/guides/faq#i-need-to-set-_id-manually-how-to-do-it-properly)

> Then the id value is always empty.

that is either because `_id` is always empty (because of the above) and / or that you are setting a custom path which would nromally be a automatic virtual

> If I extend Base, I can easily call _id or id, but any time I create a new document, TypeScript complains that _id is required!

if you do it like in the documentation, then it should not give you a error, you can also do the `implements` route, which also does not cause errors

Code used:
```ts
// NodeJS: 19.9.0
// MongoDB: 5.0 (Docker)
// Typescript 4.9.5
import { getModelForClass, prop } from '@typegoose/typegoose'; // @typegoose/typegoose@11.2.0
import { Base } from '@typegoose/typegoose/lib/defaultClasses';
import * as mongoose from 'mongoose'; // mongoose@7.2.1

// Example from the documentation via the dual interface & class way
// eslint-disable-next-line @typescript-eslint/no-empty-interface
// interface User extends Base {}
// class User {
//   @prop()
//   public username?: string;
// }

// Example via the "implements" way
class User implements Base {
  _id!: mongoose.Types.ObjectId;
  id!: string;
  @prop()
  public username?: string;
}

const UserModel = getModelForClass(User);

async function main() {
  await mongoose.connect(`mongodb://localhost:27017/`, {
    dbName: 'verifyMASTER',
  });

  await UserModel.create({ username: 'hello' });

  const doc = new UserModel({ username: 'user1' });

  console.log(doc);

  await mongoose.disconnect();
}

main();
```
(with the configs from [typegoose-testing](https://github.com/typegoose/typegoose-testing))

**Reply by theDanielJLewis** [reply]
> if you do it like in the documentation, then it should not give you a error, you can also do the implements route, which also does not cause errors

I followed https://typegoose.github.io/typegoose/docs/guides/default-classes to write:

```js
interface Something extends Base {} // have the interface to add the types of "Base" to the class
class Something extends TimeStamps {} // have your class
```

Is that not correct? What is the `implements` route? I can't find `implements` in the documentation.

**Reply by hasezoey** [reply]
both routes are correct, and both routes are shown in the code example given in https://github.com/typegoose/typegoose/discussions/851#discussioncomment-6189093, just that the one from the documentation (the "extends" route) is commented-out and the "implements route" is the non-commented-out code

see `// Example via the "implements" way` and `// Example from the documentation via the dual interface & class way`

**Reply by theDanielJLewis** [reply]
I'm sorry I'm not getting it. But I'm trying to use _both_ `Base` and `Timestamps`, and no combination or order of `implements` or `extends` in one line or two, using `class` or `interface` seems to be working for me.

**Reply by theDanielJLewis** [reply]
And even when the model definition stops giving me errors, `create()` then gives me a Mongoose error saying that _id is missing!

**Reply by hasezoey** [reply]
> create() then gives me a Mongoose error saying that _id is missing!

did you read the [documentation i linked](https://typegoose.github.io/typegoose/docs/guides/faq#i-need-to-set-_id-manually-how-to-do-it-properly) earlier?

TL;DR: dont add `@prop` if you dont want to overwrite the default implementation (which in this case you likely dont want)

> And even when the model definition stops giving me errors,

 which configuration are you using where it stops giving errors?

---

i have now put the code from https://github.com/typegoose/typegoose/discussions/851#discussioncomment-6189093 with Timestamps extension into the repository, please modify the code to show the problem you are having

reproduction repository & branch: https://github.com/typegoose/typegoose-testing/tree/verify851

**Reply by theDanielJLewis** [reply]
Thank you for your help!

I've been reading each of the links you give me. But I'm still having problems.

```js
export class MyClass extends TimeStamps implements Base {
	public _id!: Types.ObjectId;
	public id!: string;

	@prop
	public name: string;
}

// Other file

await MyModel.create(items);

// Results in "MongooseError: document must have an _id before saving"
```

**Reply by theDanielJLewis** [reply]
FACEPALM!

Typegoose 11.2 is apparently not compatible with Mongoose 7.3! I downgraded to 7.2 and I no longer get the MongooseError above.

So backing up then. What is the _right_ and _optimized_ way to implement `Base`? Is seemingly redundant code just necessary?

**Reply by theDanielJLewis** [reply]
Sorry, no. I still get problems with `_id` being required. And the above method won't let me switch it to `_id?`.
<img width="848" alt="image" src="https://github.com/typegoose/typegoose/assets/154936/daf21572-8d5f-456f-9ef2-d8bf61e80ea2">

**Reply by hasezoey** [reply]
> Typegoose 11.2 is apparently not compatible with Mongoose 7.3! I downgraded to 7.2 and I no longer get the MongooseError above.

typegoose 11.2 (currently latest) does not support mongoose as in the version range, but if explicitly installed it should work (see https://github.com/Automattic/mongoose/pull/13499#pullrequestreview-1479141291), though i will release a new version soon

> Sorry, no. I still get problems with _id being required. And the above method won't let me switch it to _id?.

i can also reproduce the error when changing to `?`, but can you provide a example of how you try to use the model where you get a `_id` required error? (also is that error a types error or a runtime error?)

**Reply by theDanielJLewis** [reply]
I looked deeper into my code and found the "problem," but I still have a preference for how it should work.

When declaring the object for `Podcast.create()`, I was making it use `PodcastClass` to help with ensuring I properly formatted the object before sending to the DB. But adding the `PodcastClass` type would result in the error telling me that `_id` was required. It's _that_ error that I've been fighting over with using `Base` but still getting typing while _creating_ my objects.

When I removed that type assignment from the unsaved object, that TS error went away.

But I would really like the ability to assign a class from Typegoose to help format the unsaved object correctly. But you can see the battle I've had with the `Base` class over this.

**Reply by hasezoey** [reply]
>  But adding the PodcastClass type would result in the error telling me that _id was required. It's that error that I've been fighting over with using Base but still getting typing while creating my objects.

for that use-case you likely will need to use `Omit` with the paths that are already existing, sadly this cannot be done (to my knowledge) at the same time simply
